core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1630,6 +1652,53 @@ private[spark] class DAGScheduler(
     }
   }
 
+  private[scheduler] def checkAndScheduleShuffleMergeFinalize(
+      shuffleStage: ShuffleMapStage): Unit = {
+    // Check if a finalize task has already been scheduled. This is to prevent the
+    // following scenario: Stage A attempt 0 fails and gets retried. Stage A attempt 1
+    // succeeded, triggering the scheduling of shuffle merge finalization. However,
+    // tasks from Stage A attempt 0 might still be running and sending task completion
+    // events to DAGScheduler. This check prevents multiple attempts to schedule merge
+    // finalization get triggered due to this.
```

Comment:
Can you double check on the comment here ?
Two things:

a) Stage A attempt 0's task will get reset when attempt 1 is launched (if non-determinate), or wont have finalize task (if determinate).
So the case described in comment should not happen ?

b) The check does handle case where we decide to finalize earlier - in case `shufflePushMinRatio < 1.0`.

Reply:
Addressed it. Let me know if this looks good or not.